File: src/navigation/ZoneChecker.cpp
#include "ZoneChecker.h"
#include <spdlog/spdlog.h>
#include <fstream>
#include <filesystem>

namespace vtol {

const std::map<std::string, double> ZoneChecker::SETTLEMENT_FALLBACK_RADII = {
    {"city", 5000.0}, {"town", 2000.0}, {"village", 700.0}, {"hamlet", 300.0}
};

ZoneChecker::ZoneChecker(const ZoneManager& zoneManager, const ZoneAvoidanceConfig& config)
    : m_zoneManager(zoneManager)
    , m_config(config)
{}

bool ZoneChecker::isNoFlyActive(const NoFlyZone& zone, double altitude) const
{
    // Per-zone override
    std::string mode = zone.avoid_mode.value_or(m_config.nofly_mode);
    if (mode == "disabled") return false;
    if (mode == "always") return true;
    // "below_altitude"
    if (zone.altitude.has_value()) return altitude < zone.altitude.value();
    return true;
}

bool ZoneChecker::isSettlementActive(double altitude) const
{
    if (m_config.settlement_mode == "disabled") return false;
    if (m_config.settlement_mode == "always") return true;
    // "below_altitude"
    return altitude < m_config.settlement_min_altitude;
}

RestrictionResult ZoneChecker::isPointRestricted(double lat, double lon, double altitude) const
{
    // Check no-fly zones
    for (const auto& zone : m_zoneManager.getAllZones()) {
        if (!isNoFlyActive(zone, altitude)) continue;
        if (nav::pointInPolygon(lat, lon, zone.points)) {
            return {true, "Запретная зона: " + zone.name};
        }
    }

    // Check settlements
    if (isSettlementActive(altitude)) {
        for (const auto& s : m_settlements) {
            // Quick bbox check
            if (lat < s.minLat || lat > s.maxLat || lon < s.minLon || lon > s.maxLon)
                continue;
            if (nav::pointInPolygon(lat, lon, s.polygon)) {
                return {true, "Населённый пункт: " + s.name + " (" + s.type + ")"};
            }
        }
    }

    return {false, ""};
}

bool ZoneChecker::segmentIntersectsObstacles(double lat1, double lon1,
                                               double lat2, double lon2, double altitude) const
{
    // Check no-fly zones
    for (const auto& zone : m_zoneManager.getAllZones()) {
        if (!isNoFlyActive(zone, altitude)) continue;
        double buffer = zone.buffer.value_or(m_config.nofly_buffer);
        Polygon buffered = buffer > 0 ? nav::polygonBuffer(zone.points, buffer) : zone.points;
        if (nav::segmentIntersectsPolygon(lat1, lon1, lat2, lon2, buffered))
            return true;
    }

    // Check settlements
    if (isSettlementActive(altitude)) {
        // Segment bbox for spatial filtering
        double minLat = std::min(lat1, lat2) - 0.05;
        double maxLat = std::max(lat1, lat2) + 0.05;
        double minLon = std::min(lon1, lon2) - 0.05;
        double maxLon = std::max(lon1, lon2) + 0.05;

        for (const auto& s : settlementsInBbox(minLat, maxLat, minLon, maxLon)) {
            Polygon buffered = m_config.settlement_buffer > 0
                ? nav::polygonBuffer(s.polygon, m_config.settlement_buffer)
                : s.polygon;
            if (nav::segmentIntersectsPolygon(lat1, lon1, lat2, lon2, buffered))
                return true;
        }
    }

    return false;
}

std::vector<Polygon> ZoneChecker::getActiveObstacles(double altitude,
                                                       double minLat, double maxLat,
                                                       double minLon, double maxLon) const
{
    std::vector<Polygon> obstacles;

    for (const auto& zone : m_zoneManager.getAllZones()) {
        if (!isNoFlyActive(zone, altitude)) continue;
        obstacles.push_back(zone.points);
    }

    if (isSettlementActive(altitude)) {
        for (const auto& s : settlementsInBbox(minLat, maxLat, minLon, maxLon)) {
            obstacles.push_back(s.polygon);
        }
    }

    return obstacles;
}

std::vector<Polygon> ZoneChecker::getBufferedObstacles(double altitude,
                                                         double minLat, double maxLat,
                                                         double minLon, double maxLon) const
{
    // Lock: this method runs in background threads (via PathPlanner),
    // while updateConfig/addSettlementFeatures modify m_config/m_settlements on main thread
    std::lock_guard lock(m_mutex);

    std::vector<Polygon> obstacles;
    int zoneCount = 0;

    for (const auto& zone : m_zoneManager.getAllZones()) {
        if (!isNoFlyActive(zone, altitude)) {
            SPDLOG_DEBUG("[ZoneChecker] Zone '{}' skipped (not active at alt={:.0f})", zone.name, altitude);
            continue;
        }
        double buffer = zone.buffer.value_or(m_config.nofly_buffer);
        auto buffered = buffer > 0 ? nav::polygonBuffer(zone.points, buffer) : zone.points;
        SPDLOG_DEBUG("[ZoneChecker] Zone '{}' included: {} pts, buffer={:.0f}m → {} buffered pts",
                     zone.name, zone.points.size(), buffer, buffered.size());
        obstacles.push_back(std::move(buffered));
        ++zoneCount;
    }

    int settlementCount = 0;
    if (isSettlementActive(altitude)) {
        for (const auto& s : settlementsInBbox(minLat, maxLat, minLon, maxLon)) {
            obstacles.push_back(m_config.settlement_buffer > 0
                ? nav::polygonBuffer(s.polygon, m_config.settlement_buffer)
                : s.polygon);
            ++settlementCount;
        }
    }

    SPDLOG_INFO("[ZoneChecker] getBufferedObstacles(alt={:.0f}): {} zones + {} settlements = {} obstacles",
                altitude, zoneCount, settlementCount, obstacles.size());
    return obstacles;
}

std::vector<Polygon> ZoneChecker::getBufferedZones(double altitude) const
{
    std::lock_guard lock(m_mutex);
    std::vector<Polygon> zones;
    for (const auto& zone : m_zoneManager.getAllZones()) {
        if (!isNoFlyActive(zone, altitude)) continue;
        double buffer = zone.buffer.value_or(m_config.nofly_buffer);
        zones.push_back(buffer > 0 ? nav::polygonBuffer(zone.points, buffer) : zone.points);
    }
    return zones;
}

std::vector<ZoneChecker::IntersectionPoint> ZoneChecker::findIntersectionPoints(
    double lat1, double lon1, double lat2, double lon2, double altitude) const
{
    std::vector<IntersectionPoint> result;

    auto findEntry = [&](const Polygon& poly, const std::string& reason) {
        // Find the first intersection of segment with polygon edges
        int n = static_cast<int>(poly.size());
        double bestT = 2.0;
        for (int i = 0; i < n; ++i) {
            int j = (i + 1) % n;
            // Segment-segment intersection parameter t on (lat1,lon1)→(lat2,lon2)
            double dx1 = lat2 - lat1, dy1 = lon2 - lon1;
            double dx2 = poly[j].lat - poly[i].lat, dy2 = poly[j].lon - poly[i].lon;
            double denom = dx1 * dy2 - dy1 * dx2;
            if (std::abs(denom) < 1e-15) continue;
            double dbx = poly[i].lat - lat1, dby = poly[i].lon - lon1;
            double t = (dbx * dy2 - dby * dx2) / denom;
            double u = (dbx * dy1 - dby * dx1) / denom;
            if (t > 0.01 && t < 0.99 && u > 0.01 && u < 0.99) {
                if (t < bestT) bestT = t;
            }
        }
        if (bestT < 2.0) {
            double iLat = lat1 + bestT * (lat2 - lat1);
            double iLon = lon1 + bestT * (lon2 - lon1);
            result.push_back({iLat, iLon, reason});
        } else {
            // Endpoint inside polygon — use midpoint as marker
            result.push_back({(lat1 + lat2) / 2.0, (lon1 + lon2) / 2.0, reason});
        }
    };

    // Check no-fly zones
    for (const auto& zone : m_zoneManager.getAllZones()) {
        if (!isNoFlyActive(zone, altitude)) continue;
        double buffer = zone.buffer.value_or(m_config.nofly_buffer);
        Polygon buffered = buffer > 0 ? nav::polygonBuffer(zone.points, buffer) : zone.points;
        if (nav::segmentIntersectsPolygon(lat1, lon1, lat2, lon2, buffered)) {
            std::string reason = "Запретная зона";
            if (!zone.name.empty()) reason += ": " + zone.name;
            findEntry(buffered, reason);
        }
    }

    // Check settlements
    if (isSettlementActive(altitude)) {
        double minLat = std::min(lat1, lat2) - 0.05;
        double maxLat = std::max(lat1, lat2) + 0.05;
        double minLon = std::min(lon1, lon2) - 0.05;
        double maxLon = std::max(lon1, lon2) + 0.05;

        for (const auto& s : settlementsInBbox(minLat, maxLat, minLon, maxLon)) {
            Polygon buffered = m_config.settlement_buffer > 0
                ? nav::polygonBuffer(s.polygon, m_config.settlement_buffer)
                : s.polygon;
            if (nav::segmentIntersectsPolygon(lat1, lon1, lat2, lon2, buffered)) {
                std::string reason = "Населённый пункт";
                if (!s.name.empty()) reason += ": " + s.name;
                findEntry(buffered, reason);
            }
        }
    }

    return result;
}

void ZoneChecker::addSettlementFeatures(const std::vector<CachedSettlement>& features)
{
    std::lock_guard lock(m_mutex);
    for (const auto& f : features) {
        m_settlements.push_back(f);
    }
    SPDLOG_DEBUG("Added {} settlement features (total: {})", features.size(), m_settlements.size());
}

void ZoneChecker::clearSettlements()
{
    std::lock_guard lock(m_mutex);
    auto count = m_settlements.size();
    m_settlements.clear();
    SPDLOG_INFO("[ZoneChecker] Cleared {} settlements", count);
}

void ZoneChecker::setSettlementCacheDir(const std::string& dir)
{
    namespace fs = std::filesystem;
    if (!fs::exists(dir)) {
        SPDLOG_WARN("Settlement cache directory not found: {}", dir);
        return;
    }

    int loaded = 0;
    for (const auto& entry : fs::directory_iterator(dir)) {
        if (entry.path().extension() != ".json") continue;
        try {
            std::ifstream file(entry.path());
            nlohmann::json data;
            file >> data;

            for (const auto& feat : data) {
                CachedSettlement s;
                std::string form = feat.value("F", "p");
                s.type = feat.value("t", "village");
                s.name = feat.value("n", "");

                if (form == "p" && feat.contains("c")) {
                    for (const auto& pt : feat["c"])
                        s.polygon.push_back({pt[0].get<double>(), pt[1].get<double>()});
                } else if (form == "c" || feat.contains("lat")) {
                    double lat = feat.value("lat", 0.0);
                    double lon = feat.value("lon", 0.0);
                    auto it = SETTLEMENT_FALLBACK_RADII.find(s.type);
                    double r = it != SETTLEMENT_FALLBACK_RADII.end() ? it->second : 500.0;
                    s.polygon = nav::circleToPolygon(lat, lon, r, 16);
                }

                if (s.polygon.size() < 3) continue;

                // Precompute bbox
                s.minLat = s.maxLat = s.polygon[0].lat;
                s.minLon = s.maxLon = s.polygon[0].lon;
                for (const auto& p : s.polygon) {
                    s.minLat = std::min(s.minLat, p.lat);
                    s.maxLat = std::max(s.maxLat, p.lat);
                    s.minLon = std::min(s.minLon, p.lon);
                    s.maxLon = std::max(s.maxLon, p.lon);
                }

                m_settlements.push_back(std::move(s));
                ++loaded;
            }
        } catch (const std::exception& e) {
            SPDLOG_WARN("Failed to load settlement file {}: {}", entry.path().string(), e.what());
        }
    }

    SPDLOG_INFO("Loaded {} settlements from cache dir {}", loaded, dir);
}

std::vector<CachedSettlement> ZoneChecker::settlementsInBbox(double minLat, double maxLat,
                                                               double minLon, double maxLon) const
{
    std::vector<CachedSettlement> result;
    for (const auto& s : m_settlements) {
        if (s.maxLat < minLat || s.minLat > maxLat ||
            s.maxLon < minLon || s.minLon > maxLon)
            continue;
        result.push_back(s);
    }
    return result;
}

} // namespace vtol

File: src/navigation/zone_checker.py
import json
import logging
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Tuple

from src.core.config import ZoneAvoidanceConfig
from src.navigation.zone_manager import ZoneManager
from src.navigation.calculations import (
    point_in_polygon, polygon_buffer, segment_intersects_polygon,
    circle_to_polygon, haversine_distance, meters_to_lat_offset, meters_to_lon_offset,
    along_track_fraction, cross_track_distance,
)

logger = logging.getLogger(__name__)

_SETTLEMENT_FALLBACK_RADII: Dict[str, float] = {
    'city': 5000.0,
    'town': 2000.0,
    'village': 700.0,
    'hamlet': 300.0,
}


@dataclass
class ObstaclePolygon:
    points: List[List[float]]   # [[lat, lon], ...]
    buffer: float               # meters
    source: str                 # "nofly" | "settlement"
    name: str
    min_altitude: Optional[float]  # None = restricted at all altitudes


@dataclass
class _CachedSettlement:
    """Settlement feature with precomputed bounding box for spatial filtering."""
    feat: dict
    min_lat: float
    max_lat: float
    min_lon: float
    max_lon: float


class ZoneChecker:
    def __init__(self, zone_manager: ZoneManager, config: ZoneAvoidanceConfig):
        self._zone_manager = zone_manager
        self._config = config
        self._settlements: List[_CachedSettlement] = []
        self._cache_dir: Optional[Path] = None

    def set_config(self, config: ZoneAvoidanceConfig):
        self._config = config

    def set_settlement_cache_dir(self, cache_dir: Path):
        self._cache_dir = cache_dir
        self._load_settlement_cache()

    def _load_settlement_cache(self):
        self._settlements = []
        if not self._cache_dir or not self._cache_dir.exists():
            return
        try:
            for f in self._cache_dir.glob('*.json'):
                with open(f, 'r', encoding='utf-8') as fh:
                    features = json.load(fh)
                for feat in features:
                    cached = self._make_cached_settlement(feat)
                    if cached:
                        self._settlements.append(cached)
        except Exception as e:
            logger.warning("Failed to load settlement cache: %s", e)

    def _make_cached_settlement(self, feat: dict) -> Optional[_CachedSettlement]:
        """Precompute bounding box for fast spatial filtering."""
        if feat.get('F') == 'p' and 'c' in feat:
            coords = feat['c']
            if len(coords) < 3:
                return None
            lats = [c[0] for c in coords]
            lons = [c[1] for c in coords]
            return _CachedSettlement(feat, min(lats), max(lats), min(lons), max(lons))
        elif feat.get('F') == 'n':
            lat = feat.get('lat')
            lon = feat.get('lon')
            if lat is None or lon is None:
                return None
            stype = feat.get('t', 'hamlet')
            radius = _SETTLEMENT_FALLBACK_RADII.get(stype, 500.0)
            dlat = meters_to_lat_offset(radius)
            dlon = meters_to_lon_offset(radius, lat)
            return _CachedSettlement(feat, lat - dlat, lat + dlat, lon - dlon, lon + dlon)
        return None

    def reload_settlements(self):
        self._load_settlement_cache()

    def add_settlement_features(self, features: list):
        """Incrementally add new settlement features (from freshly fetched tiles)."""
        added = 0
        for feat in features:
            cached = self._make_cached_settlement(feat)
            if cached:
                self._settlements.append(cached)
                added += 1
        if added:
            logger.debug("[ZoneChecker] Added %d settlements incrementally (total: %d)",
                         added, len(self._settlements))

    def _get_settlement_polygon(self, feat: dict) -> Optional[List[List[float]]]:
        """Extract polygon from settlement feature."""
        if feat.get('F') == 'p' and 'c' in feat:
            return feat['c']
        elif feat.get('F') == 'n':
            lat = feat.get('lat')
            lon = feat.get('lon')
            stype = feat.get('t', 'hamlet')
            if lat is not None and lon is not None:
                radius = _SETTLEMENT_FALLBACK_RADII.get(stype, 500.0)
                return circle_to_polygon(lat, lon, radius, 16)
        return None

    def _settlements_in_bbox(self, min_lat: float, max_lat: float,
                             min_lon: float, max_lon: float) -> List[_CachedSettlement]:
        """Fast bbox filter — returns only settlements that overlap the query box."""
        result = []
        for s in self._settlements:
            if s.max_lat < min_lat or s.min_lat > max_lat:
                continue
            if s.max_lon < min_lon or s.min_lon > max_lon:
                continue
            result.append(s)
        return result

    def is_point_restricted(self, lat: float, lon: float,
                            altitude: float) -> Tuple[bool, Optional[str]]:
        """Check if point is in a restricted area.
        Returns (restricted, reason_string)."""
        # Check nofly zones
        for zone in self._zone_manager.get_all_zones():
            mode = zone.avoid_mode if zone.avoid_mode else self._config.nofly_mode
            if mode == "disabled":
                continue

            buf = zone.buffer if zone.buffer is not None else self._config.nofly_buffer

            if mode == "below_altitude" and zone.altitude is not None:
                if altitude >= zone.altitude:
                    continue

            poly = zone.points
            if buf > 0:
                poly = polygon_buffer(poly, buf)
            if point_in_polygon(lat, lon, poly):
                reason = f"Запретная зона: {zone.name or zone.id}"
                if mode == "below_altitude" and zone.altitude:
                    reason += f" (ниже {zone.altitude:.0f} м)"
                return True, reason

        # Check settlements — bbox filter first
        if self._config.settlement_mode != "disabled":
            if self._config.settlement_mode == "below_altitude":
                if altitude >= self._config.settlement_min_altitude:
                    return False, None

            # Expand search box by buffer
            buf = self._config.settlement_buffer
            dlat = meters_to_lat_offset(buf) if buf > 0 else 0.01
            dlon = meters_to_lon_offset(buf, lat) if buf > 0 else 0.01
            nearby = self._settlements_in_bbox(lat - dlat, lat + dlat, lon - dlon, lon + dlon)

            for cached in nearby:
                poly = self._get_settlement_polygon(cached.feat)
                if poly is None:
                    continue
                buf_poly = poly
                if self._config.settlement_buffer > 0:
                    buf_poly = polygon_buffer(poly, self._config.settlement_buffer)
                if point_in_polygon(lat, lon, buf_poly):
                    stype = cached.feat.get('t', 'населённый пункт')
                    reason = f"Населённый пункт ({stype})"
                    if self._config.settlement_mode == "below_altitude":
                        reason += f" (ниже {self._config.settlement_min_altitude:.0f} м)"
                    return True, reason

        return False, None

    def segment_intersects_obstacles(self, lat1: float, lon1: float,
                                     lat2: float, lon2: float,
                                     altitude: float) -> bool:
        """Check if segment crosses any active obstacle.
        Uses bbox prefilter for settlements."""
        # NoFly zones (few, check all)
        for zone in self._zone_manager.get_all_zones():
            mode = zone.avoid_mode if zone.avoid_mode else self._config.nofly_mode
            if mode == "disabled":
                continue
            if mode == "below_altitude" and zone.altitude is not None:
                if altitude >= zone.altitude:
                    continue
            buf = zone.buffer if zone.buffer is not None else self._config.nofly_buffer
            poly = zone.points
            if buf > 0:
                poly = polygon_buffer(poly, buf)
            if segment_intersects_polygon(lat1, lon1, lat2, lon2, poly):
                return True

        # Settlements — bbox filter
        if self._config.settlement_mode != "disabled":
            if self._config.settlement_mode == "below_altitude":
                if altitude >= self._config.settlement_min_altitude:
                    return False

            buf = self._config.settlement_buffer
            min_lat = min(lat1, lat2) - meters_to_lat_offset(buf + 1000)
            max_lat = max(lat1, lat2) + meters_to_lat_offset(buf + 1000)
            mid_lat = (lat1 + lat2) / 2
            min_lon = min(lon1, lon2) - meters_to_lon_offset(buf + 1000, mid_lat)
            max_lon = max(lon1, lon2) + meters_to_lon_offset(buf + 1000, mid_lat)

            nearby = self._settlements_in_bbox(min_lat, max_lat, min_lon, max_lon)
            for cached in nearby:
                poly = self._get_settlement_polygon(cached.feat)
                if poly is None:
                    continue
                if self._config.settlement_buffer > 0:
                    poly = polygon_buffer(poly, self._config.settlement_buffer)
                if segment_intersects_polygon(lat1, lon1, lat2, lon2, poly):
                    return True

        return False

    def get_active_obstacles(self, altitude: float,
                             bbox: Optional[Tuple[float, float, float, float]] = None
                             ) -> List[ObstaclePolygon]:
        """Collect active obstacle polygons for given altitude.
        bbox: optional (min_lat, max_lat, min_lon, max_lon) for spatial filtering."""
        obstacles = []

        # NoFly zones (always all — typically few)
        for zone in self._zone_manager.get_all_zones():
            mode = zone.avoid_mode if zone.avoid_mode else self._config.nofly_mode
            if mode == "disabled":
                continue
            if mode == "below_altitude" and zone.altitude is not None:
                if altitude >= zone.altitude:
                    continue

            buf = zone.buffer if zone.buffer is not None else self._config.nofly_buffer
            obstacles.append(ObstaclePolygon(
                points=zone.points,
                buffer=buf,
                source="nofly",
                name=zone.name or zone.id,
                min_altitude=zone.altitude if mode == "below_altitude" else None,
            ))

        # Settlements — spatial filter required
        if self._config.settlement_mode != "disabled":
            if self._config.settlement_mode == "below_altitude":
                if altitude >= self._config.settlement_min_altitude:
                    return obstacles

            if bbox:
                nearby = self._settlements_in_bbox(*bbox)
            else:
                nearby = self._settlements  # no bbox = all (legacy fallback)

            min_alt = self._config.settlement_min_altitude if self._config.settlement_mode == "below_altitude" else None
            for cached in nearby:
                poly = self._get_settlement_polygon(cached.feat)
                if poly is None:
                    continue
                stype = cached.feat.get('t', 'settlement')
                obstacles.append(ObstaclePolygon(
                    points=poly,
                    buffer=self._config.settlement_buffer,
                    source="settlement",
                    name=stype,
                    min_altitude=min_alt,
                ))

        return obstacles

    def get_buffered_obstacles(self, altitude: float,
                               bbox: Optional[Tuple[float, float, float, float]] = None
                               ) -> List[List[List[float]]]:
        """Get buffered polygons for visibility graph / intersection checks."""
        result = []
        for obs in self.get_active_obstacles(altitude, bbox=bbox):
            if obs.buffer > 0:
                result.append(polygon_buffer(obs.points, obs.buffer))
            else:
                result.append(obs.points)
        return result

    def segment_intersects_obstacles_climb(self, lat1: float, lon1: float,
                                            lat2: float, lon2: float,
                                            start_alt: float, end_alt: float) -> bool:
        """Check segment intersection with altitude interpolation for climb_enroute.
        For below_altitude zones, interpolate altitude along the segment and skip
        zones where the aircraft is above the zone limit at that point."""
        # NoFly zones
        for zone in self._zone_manager.get_all_zones():
            mode = zone.avoid_mode if zone.avoid_mode else self._config.nofly_mode
            if mode == "disabled":
                continue

            buf = zone.buffer if zone.buffer is not None else self._config.nofly_buffer
            poly = zone.points
            if buf > 0:
                poly = polygon_buffer(poly, buf)

            if mode == "below_altitude" and zone.altitude is not None:
                # Interpolate altitude at zone centroid position along segment
                cx = sum(v[0] for v in poly) / len(poly)
                cy = sum(v[1] for v in poly) / len(poly)
                frac = along_track_fraction(cx, cy, lat1, lon1, lat2, lon2)
                alt_at_zone = start_alt + (end_alt - start_alt) * frac
                if alt_at_zone >= zone.altitude:
                    continue

            if segment_intersects_polygon(lat1, lon1, lat2, lon2, poly):
                return True

        # Settlements
        if self._config.settlement_mode != "disabled":
            # For climb: check with minimum altitude (most conservative for settlements)
            min_alt = min(start_alt, end_alt)
            if self._config.settlement_mode == "below_altitude":
                if min_alt >= self._config.settlement_min_altitude:
                    return False

            buf = self._config.settlement_buffer
            min_lat = min(lat1, lat2) - meters_to_lat_offset(buf + 1000)
            max_lat = max(lat1, lat2) + meters_to_lat_offset(buf + 1000)
            mid_lat = (lat1 + lat2) / 2
            min_lon = min(lon1, lon2) - meters_to_lon_offset(buf + 1000, mid_lat)
            max_lon = max(lon1, lon2) + meters_to_lon_offset(buf + 1000, mid_lat)

            nearby = self._settlements_in_bbox(min_lat, max_lat, min_lon, max_lon)
            for cached in nearby:
                poly = self._get_settlement_polygon(cached.feat)
                if poly is None:
                    continue
                if self._config.settlement_buffer > 0:
                    poly = polygon_buffer(poly, self._config.settlement_buffer)

                if self._config.settlement_mode == "below_altitude":
                    cx = sum(v[0] for v in poly) / len(poly)
                    cy = sum(v[1] for v in poly) / len(poly)
                    frac = along_track_fraction(cx, cy, lat1, lon1, lat2, lon2)
                    alt_at_settlement = start_alt + (end_alt - start_alt) * frac
                    if alt_at_settlement >= self._config.settlement_min_altitude:
                        continue

                if segment_intersects_polygon(lat1, lon1, lat2, lon2, poly):
                    return True

        return False

    def get_buffered_obstacles_climb(self, start_alt: float, end_alt: float,
                                      bbox: Optional[Tuple[float, float, float, float]] = None,
                                      start_lat: float = 0, start_lon: float = 0,
                                      end_lat: float = 0, end_lon: float = 0
                                      ) -> List[List[List[float]]]:
        """Get buffered obstacles with altitude profile — skip zones the aircraft
        will fly over at that segment position."""
        result = []

        for zone in self._zone_manager.get_all_zones():
            mode = zone.avoid_mode if zone.avoid_mode else self._config.nofly_mode
            if mode == "disabled":
                continue

            if mode == "below_altitude" and zone.altitude is not None:
                cx = sum(v[0] for v in zone.points) / len(zone.points)
                cy = sum(v[1] for v in zone.points) / len(zone.points)
                frac = along_track_fraction(cx, cy, start_lat, start_lon, end_lat, end_lon)
                alt_at_zone = start_alt + (end_alt - start_alt) * frac
                if alt_at_zone >= zone.altitude:
                    continue

            buf = zone.buffer if zone.buffer is not None else self._config.nofly_buffer
            if buf > 0:
                result.append(polygon_buffer(zone.points, buf))
            else:
                result.append(zone.points)

        # Settlements
        if self._config.settlement_mode != "disabled":
            min_alt = min(start_alt, end_alt)
            if self._config.settlement_mode == "below_altitude":
                if min_alt >= self._config.settlement_min_altitude:
                    return result

            if bbox:
                nearby = self._settlements_in_bbox(*bbox)
            else:
                nearby = self._settlements

            for cached in nearby:
                poly = self._get_settlement_polygon(cached.feat)
                if poly is None:
                    continue

                if self._config.settlement_mode == "below_altitude":
                    cx = sum(v[0] for v in poly) / len(poly)
                    cy = sum(v[1] for v in poly) / len(poly)
                    frac = along_track_fraction(cx, cy, start_lat, start_lon, end_lat, end_lon)
                    alt_at_settlement = start_alt + (end_alt - start_alt) * frac
                    if alt_at_settlement >= self._config.settlement_min_altitude:
                        continue

                if self._config.settlement_buffer > 0:
                    result.append(polygon_buffer(poly, self._config.settlement_buffer))
                else:
                    result.append(poly)

        return result
